src/families.h

```diff
@@ -18,139 +18,106 @@
-        + alpha_i;
-      double mu_n = 2.0*n_samples*alpha_i;
-      step_sizes.push_back(1.0 / (2.0*L + std::min(L, mu_n)));
-    }
-    return step_sizes;
-  }
+  Family(const double L_scaling) : L_scaling(L_scaling) {};
+
+  void Preprocess(std::vector<double>& y,
+                  std::vector<double>& y_center,
+                  std::vector<double>& y_scale) noexcept;
+
+  double Loss(const std::vector<double>& prediction,
+              const std::vector<double>& y,
+              const unsigned             i) const noexcept;
+
+  void Gradient(const std::vector<double>& prediction,
+                const std::vector<double>& y,
+                const unsigned             i,
+                std::vector<double>&       gradient) const noexcept;
+
+  double NullDeviance(const std::vector<double>& y,
+                      const bool                 fit_intercept,
+                      const unsigned             n_classes) const noexcept;
+
+  Eigen::MatrixXd
+  LambdaResponse(const std::vector<double>& y,
+                 std::vector<double>&       y_mat_scale) const noexcept;
 
-  std::vector<double> y;
-  std::vector<double> y_center;
-  std::vector<double> y_scale;
-  Eigen::MatrixXd y_mat;
-  std::vector<double> y_mat_scale;
-  unsigned n_samples;
-  unsigned n_classes;
-  unsigned n_targets;
-  double null_deviance = 0.0;
   double L_scaling;
-  double lambda_scaling = 1.0;
 };
 
 class Gaussian : public Family {
 public:
-  Gaussian(const std::vector<double>& y,
-           const unsigned n_samples,
-           const unsigned n_classes)
-           : Family(y, n_samples, n_classes, 1.0) {}
+  Gaussian() : Family(1.0) {}
 
-  void PreprocessResponse() {
+  void Preprocess(std::vector<double>& y,
+                  std::vector<double>& y_center,
+                  std::vector<double>& y_scale) noexcept {
     y_center[0] = Mean(y);
     y_scale[0] = StandardDeviation(y, y_center[0]);
-    y_mat_scale[0] = y_scale[0];
-
-    for (std::size_t i = 0; i < n_samples; ++i) {
-      y[i] -= y_center[0];
-      y[i] /= y_scale[0];
-      y_mat(i, 0) = y[i];
-    }
 
-    lambda_scaling = y_scale[0];
+    for (auto& y_i : y)
+      y_i = (y_i - y_center[0])/y_scale[0];
   }
 
-  double Link(const unsigned i) { return y[i]; }
-
-  double Loss(const double prediction, const unsigned i) {
-    return 0.5*(prediction - y[i])*(prediction - y[i]);
+  double Loss(const std::vector<double>& prediction,
+              const std::vector<double>& y,
+              const unsigned             i) const noexcept {
+    return 0.5*(prediction[0] - y[i])*(prediction[0] - y[i]);
   }
 
-  double Loss(const std::vector<double>& prediction, const unsigned i) {
-    return Loss(prediction[0], i);
+  void Gradient(const std::vector<double>& prediction,
+                const std::vector<double>& y,
+                const unsigned             i,
+                std::vector<double>&       gradient) const noexcept {
+    gradient[0] = prediction[0] - y[i];
```

Comment:
I'm a bit confused on this interface - why are we passing (even by reference) a whole vector of `y` and then dereferencing internally to element `i`? Wouldn't it be easier to just have `const double y_obs` as an argument?

Reply:
These functions are implicitly inlined, so I am counting on the compiler to work this out without any loss of performance. The real reason, however, is that we need the entire vector for the multivariate families so can't pass in a double.